File: CSCD 240/Lab 12/cscd240lab12/LinkedList.c
//Dan Herve
//CSCD 240: Lab 12

#include "list.h"

LinkedList * linkedList(){ //Make a LinkedList

    return (LinkedList*)calloc(1, sizeof(LinkedList));
}

void clear(LinkedList * myList){ //free the Nodes

    Node *curr;

    if(myList->size == 0);
    else {

        for(curr = myList->head; curr != NULL; curr = myList->head){

            myList->head = myList->head->next; //chop off the first node
            free(curr);
            myList->size--;
        }
    }
}

int size(LinkedList * myList){ //return LinkedList size

    if(myList == NULL)
        return 0;

    return myList->size;
}

void addFirst(LinkedList * myList, int d){//add Node to beginning of list

    Node *node = (Node*)calloc(1, sizeof(Node));

    node->data = d;

    node->next = myList->head;
    myList->head = node;
    myList->size++;
}

void addLast(LinkedList * myList, int d){//add Node to end of list

    Node *node = (Node*)calloc(1, sizeof(Node));
    node->data = d;
    Node *curr;

    if(myList->size == 0) {

        myList->head = node;
        myList->size++;
    }

    for(curr = myList->head; curr->next != NULL; curr = curr->next);

    curr->next = node;
    myList->size++;
}

int addDataByIndex(LinkedList * myList, int d, int index){

    Node *node = (Node*)calloc(1, sizeof(Node));//new Node
    node->data = d;
    int count = 0;
    Node *curr = myList->head->next;
    Node *prev = myList->head;

    if((myList->size == 0 && index == 0) || index == 0) {

        free(node);//wasn't used
        addFirst(myList, d);
        return 1;
    }
    else if (index - 1 >= myList->size) {

        free(node);//wasn't used
        printf("Index out of Bounds!\n");
        return 0;
    }

    for(curr = myList->head; count < index; prev = curr, curr = curr->next, count++);

    node->next = curr;
    prev->next = node;
    myList->size++;
    return 1;
}

void sort(LinkedList * myList){//selection sort

    Node *curr;
    Node *prev;
    int smallest = 0;
    int temp = 0;

    if(myList->size == 0);

    for(prev = myList->head; prev != NULL; prev = prev->next) {

        smallest = prev->data;

        for(curr = prev->next; curr != NULL; curr = curr->next) {

            if(smallest > curr->data) {

                temp = smallest;
                smallest = curr->data;
                prev->data = smallest;
                curr->data = temp;
            }
        }
    }
}

Node * removeByIndex(LinkedList * theList, int index){

    int count = 0;
    Node *curr = theList->head;
    Node *prev = theList->head;

    if(theList->size == 0) {

        return NULL;
    }

    if(index - 1 >= theList->size) {

        printf("Index out of Bounds!\n");
        return NULL;
    }

    for(curr = theList->head; count < index; prev = curr, curr = curr->next, count++);

    if(index == 0)//have to move head if removing head Node
        theList->head = curr->next;
    prev->next = curr->next;
    theList->size--;
    return curr;
}

int removeByValue(LinkedList * theList, int d) {

    Node *curr;
    Node *prev = theList->head;
    int data = 0;

    if(theList->size == 0) {

        return 0;
    }

    for(curr = theList->head; curr != NULL; prev = curr, curr = curr->next) {

        if(curr->data == d) {

            if(curr == theList->head)//have to move head if removing head Node
                theList->head = curr->next;
            prev->next = curr->next;
            theList->size--;
            data = curr->data;
            free(curr);
            return data;
        }
    }

    return 0;
}

void printList(LinkedList * myList) {

    if(myList == NULL || myList->head == NULL)
        printf("Empty List\n");
    else {

        Node *curr;
        printf("[");
        for (curr = myList->head; curr != NULL; curr = curr->next){

            if(curr->next == NULL)
                printf("%d]", curr->data);
            else
                printf("%d, ", curr->data);
        }
        printf("\n");
    }
}

File: CSCD 300 Data Structures/Lab 7/LinkedList.java
public class LinkedList {

   private class Node implements Comparable<Node>{

      private Comparable data;
      private Node next;
      
      public Node(Comparable data) {
      
         this.data = data;
      }
   
      public Node(Comparable data, Node next) {
   
         this.data = data;
         this.next = next;
      }
   
      @Override
      public int compareTo(Node that) {
   
         return this.data.compareTo(that.data);
      }
   
      @Override
      public String toString() {
   
         return "" + data;
      }
   }

   Node head = new Node(null);
   int size = 0;
   
   public int getSize() {
   
      return size;
   }
   
   public void addFirst(Comparable data) {
      
      Node temp = new Node(data, head.next);
      head.next = temp;    
      size++;
   }
   
   public void addOrdered(Comparable data) {
   
      if (size == 0)
      
         addFirst(data);
         
      else {
         
         Node curr;
         Node prev = head;
         Node temp = new Node(data);
         
         for (curr = head.next; curr != null; prev = curr, curr = curr.next) {
         
            if(curr.data.compareTo(data) == 0) { //replaces duplicates
         
               temp.next = curr.next;
               prev.next = temp;
               size++;
               break;
            }
         
            else if (curr.data.compareTo(data) > 0) {
            
               temp.next = curr;
               prev.next = temp;
               size++;
               break;
            }
         }
         
         if(curr == null) { //if inserted value is higher than the rest
         
            temp.next = prev.next;
            prev.next = temp;
            size++;
         }
      }
   }
   
   public void removeFirst() {
      
      head.next = head.next.next; 
      size--;
   }
   
   public Comparable removeOrdered(Comparable data) {
   
      if (size == 0)
      
         return null;
         
      else {
         
         Node curr;
         Node prev = head;
         Node temp;
         
         for (curr = head.next; curr != null; prev = curr, curr = curr.next) {
         
            if (curr.data.compareTo(data) == 0) {
               
               temp = curr;
               prev.next = curr.next;
               curr = null;
               size--;
               return temp.data;
            }
         }
         
         return null;
      }   
   }
   
   public Comparable getData(Comparable target) {
   
      Node curr;
      
      for(curr = head.next; curr != null; curr = curr.next) {
      
         if (curr.data.compareTo(target) == 0) {
         
            return curr.data;
         }
      }
         
      return null;
   }
   
   public void printList() {
   
      Node curr;  
   
      for(curr = head.next; curr != null; curr = curr.next) {
      
         if(curr.next == null) 
            System.out.print(curr.data.toString());
         else   
            System.out.print(curr.data.toString()+ " ");
      }
   }
   
   public void clear() {
   
      head.next = null;
      size = 0;
   }   
}